fem/dgmassinv.hpp

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_DGMASSINV_HPP
+#define MFEM_DGMASSINV_HPP
+
+#include "../linalg/operator.hpp"
+#include "fespace.hpp"
+
+namespace mfem
+{
+
+/// @brief Solver for the discontinuous Galerkin mass matrix.
+///
+/// This class performs a @a local (diagonally preconditioned) conjugate
+/// gradient iteration for each element. Optionally, a change of basis is
+/// performed to iterate on a better-conditioned system. This class fully
+/// supports execution on device (GPU).
+class DGMassInverse : public Solver
+{
+protected:
+   DG_FECollection fec; ///< FE collection in requested basis.
+   FiniteElementSpace fes; ///< FE space in requested basis.
+   const DofToQuad *d2q; ///< Change of basis. Not owned.
+   Array<double> B_; ///< Inverse of change of basis.
+   Array<double> Bt_; ///< Inverse of change of basis, transposed.
+   class BilinearForm *M; ///< Mass bilinear form, owned.
+   class MassIntegrator *m; ///< Mass integrator, owned by the form @ref M.
+   Vector diag_inv; ///< Jacobi preconditioner.
+   double rel_tol = 1e-12; ///< Relative CG tolerance.
+   double abs_tol = 1e-12; ///< Absolute CG tolerance.
+   int max_iter = 100; ///< Maximum number of CG iterations;
+
+   /// @name Intermediate vectors needed for CG three-term recurrence.
+   ///@{
+   mutable Vector r_, d_, z_, b2_;
```

Comment:
It's just a suggestion/open-question, but would it be possible to use the `_` only when needed (like in the body of `SetRelTol` and `SetAbsTol`) ? 
Especially for header files, I have the impression of them being more readable without such extra characters.

Reply:
The reason for these is because later I have something like
```c++
auto r = r_.Write();
auto d = d_.Write();
auto z = z_.Write();
auto u = u_.ReadWrite();
```
I could change this to have the `d` prefix, e.g. `d_r` for "device pointer"… I don't know what our general convention should be here.